src/Features/LanguageServer/Protocol/Extensions/ProtocolConversions.cs

```diff
@@ -79,19 +83,53 @@ public static Completion.CompletionTrigger LSPToRoslynCompletionTrigger(LSP.Comp
             }
             else if (context.TriggerKind == LSP.CompletionTriggerKind.Invoked)
             {
-                return Completion.CompletionTrigger.Invoke;
+                if (context is not LSP.VSCompletionContext vsCompletionContext)
+                {
+                    return Completion.CompletionTrigger.Invoke;
+                }
+
+                switch (vsCompletionContext.InvokeKind)
+                {
+                    case LSP.VSCompletionInvokeKind.Explicit:
+                        return Completion.CompletionTrigger.Invoke;
+
+                    case LSP.VSCompletionInvokeKind.Typing:
+                        var insertionChar = await GetInsertionCharacter(document, position, cancellationToken).ConfigureAwait(false);
+                        return Completion.CompletionTrigger.CreateInsertionTrigger(insertionChar);
+
+                    case LSP.VSCompletionInvokeKind.Deletion:
+                        Contract.ThrowIfNull(context.TriggerCharacter);
+                        Contract.ThrowIfFalse(char.TryParse(context.TriggerCharacter, out var triggerChar));
+                        return Completion.CompletionTrigger.CreateDeletionTrigger(triggerChar);
+
+                    default:
+                        // LSP added an InvokeKind that we need to support.
+                        Logger.Log(FunctionId.LSPCompletion_MissingLSPCompletionInvokeKind);
```

Comment:
Sorry for all the questions, just trying to ramp a bit more on what Roslyn does. What does logging an ID like this do?

Reply:
I think this is for telemetry when executed in VS

---
In reply to: [570453474](https://github.com/dotnet/roslyn/pull/51002#discussion_r570453474) [](ancestors = 570453474)

Reply:
Hmm, so it's more like failure recognition telemetry? Is there a process you typically follow to follow up on telemetry bits like this to ensure failures don't go unseen?

Reply:
We monitors NFW (or at least we have queries/dashboard for them, not sure if it's monitored though), but not telemetry like this. It'd be up to feature owners to decide how to track them.

Reply:
1. i'm not a fan.  i'm not sure i'm aware of us *ever* checking these functionids for any purpose.  i think we need a rule that if we create a functionid that someone is assigend to track it, and they need to have a recurring task to check in and see if something is wrong.  otherwise it's just more noise in the noisiest system ever.

Reply:
Is there an alternative approach people would recommend? I was thinking we could throw here instead of sending a telemetry event, but it would break the completion experience for users in the meantime while we responded to the new invoke and/or trigger kind.

Reply:
So I just left it as is for now since while the telemetry approach isn't ideal, there doesn't seem to be a better alternative.